src/Balances.sol

```diff
@@ -26,16 +26,28 @@ contract Balances {
         return scheduledForTransfer.length;
     }
 
+    function participantIsReadyForTransfer (address participant) public view returns (bool) {
+        for (uint i = 0; i < readyForTransfer.length; i++) {
+            if (readyForTransfer[i] == participant) {
+                return true;
+            }
+        }
+        return false;
+    }
+
     function increaseParticipantBalance(
         address payable participant,
         uint amount
     ) internal {
         uint oldBalance = balances[participant];
         uint newBalance = oldBalance + amount;
         balances[participant] = newBalance;
+        if (newBalance <= minBalanceForTransfer) {
+            return;
+        }
         if (
-            oldBalance <= minBalanceForTransfer &&
-            newBalance > minBalanceForTransfer
+            oldBalance <= minBalanceForTransfer
+            || !participantIsReadyForTransfer(participant)
```

Comment:
Is it safe to skip the check `!participantIsReadyForTransfer(participant)`? I understand that `oldBalance <= minBalanceForTransfer` costs less gas than `participantIsReadyForTransfer`, which is a good optimisation.

I am struggling to understand if this is correct, though.

Situations to consider:
- minBalanceForTransfer is increased
- minBalanceForTransfer is decreased

Reply:
### If the min balance for transfer is increased
- participants that were below the threshold
  - nothing happens, as they are now even further below the threshold
- participants that were above the threshold before
  - and will be above the threshold after
    - they will stay in the array and not be added twice, as the situation is as if the threshold was never changed
  - and will be below the threshold after
    - they are kept in the `readyForTransfer` structure, and not removed. I think that's ok. They won't be added twice either  (their old balance can't be below the threshold if they were already above it, and they are in the array of readyToTransfer participants)

### If the min balance for transfer is decreased
- participants that were above the threshold
  - will stay in `readyForTransfer`, and not be removed (this code path doesn't exist)
  - will not be added twice, as
    - `oldbalance <= minBalanceForTransfer` is false (they were already above the previous higher threshold)
    - as is `!participantIsReadyForTransfer(..)` (they are already in the array)
- participants that were below the threshold
  - if they are still below the threshold, nothing happens
  - if they are now above the threshold, they are marked as ready

Reply:
I'll need to think about this a bit.

In the meantime, what do you think about adding tests for the different situations you described above? This would ensure that the implementation handles them exactly as it should.

Reply:
all covered in passing tests now, no contract changes necessary

Reply:
> participants that were above the threshold before
> and will be below the threshold after
> they are kept in the readyForTransfer structure, and not removed. I think that's ok.

This means that we will pay rewards to these participants on the next payout cycle even though they don't meet the new threshold requirement.

I don't have a strong opinion about whether this is ok or not; I am happy to follow your decision.

I think it's worth adding a code comment to explain that this behaviour is intentional.

Reply:
The comment is there in tests, to me that's good enough - if one would change the logic, this test would start failing. Wdyt?

Reply:
We can fix this if it becomes a problem